File: analysis/pipeline_demo_broadened.py
"""
pipeline_demo_broadened.py

PIPELINE DEMONSTRATION ONLY — NOT A SCIENTIFIC RESULT.

Runs the full downstream statistical pipeline (propensity score model, IPTW
weighting, balance diagnostics, weighted Cox model, weighted KM curve)
against a BROADENED cohort from the open MIMIC-IV Demo (100 patients):
all sepsis ICU stays, split by "received any vasopressor during the stay"
vs. "did not" — NOT the strict Cryptic Shock definition in PROTOCOL.md.

Why broadened: the strict Cryptic Shock T0 definition (see
scripts/build_cohort_demo.py) found only 1 eligible patient in the 100-patient
demo — too few for any weighting/regression method to run meaningfully. This
script exists purely to build and debug the code for every remaining phase
(Sections 4.1-4.4 of PROTOCOL.md) so it's ready to point at the real,
full-scale MIMIC-IV cohort once PhysioNet credentialing is approved.

Do not cite, report, or present any number this script outputs as a finding.
Sample size (~20-something patients) is far too small for a valid estimate.

Requires: duckdb, pandas, numpy, statsmodels, lifelines, matplotlib
"""

import duckdb
import numpy as np
import pandas as pd
import statsmodels.api as sm
from lifelines import CoxPHFitter, KaplanMeierFitter
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "mimic-iv-clinical-database-demo-2.2"
OUT_DIR = ROOT / "data" / "processed"
REPORT_DIR = ROOT / "report"
OUT_DIR.mkdir(parents=True, exist_ok=True)
REPORT_DIR.mkdir(parents=True, exist_ok=True)

LACTATE_ITEMIDS = (50813, 52442, 53154)
MAP_ITEMIDS = (220052, 220181)
VASOPRESSOR_ITEMIDS = (221906, 221289, 222315, 221749, 221662)
FOLLOWUP_DAYS = 28

SEPSIS_ICD9 = ("0380", "0381", "03810", "03811", "03812", "03819", "0382", "0383",
               "03840", "03841", "03842", "03843", "03844", "03849", "0388", "0389",
               "99591", "99592", "78552")
SEPSIS_ICD10_PREFIXES = ("A40", "A41", "R6520", "R6521")


def csv(relpath: str) -> str:
    return str(DATA / relpath).replace("\\", "/")


# ---------------------------------------------------------------------------
# 1. Build the broadened cohort
# ---------------------------------------------------------------------------

def build_cohort(con):
    patients = con.execute(f"select * from read_csv_auto('{csv('hosp/patients.csv.gz')}')").df()
    admissions = con.execute(f"select * from read_csv_auto('{csv('hosp/admissions.csv.gz')}')").df()
    icustays = con.execute(f"select * from read_csv_auto('{csv('icu/icustays.csv.gz')}')").df()
    diagnoses = con.execute(f"select * from read_csv_auto('{csv('hosp/diagnoses_icd.csv.gz')}')").df()

    diagnoses["icd_code"] = diagnoses["icd_code"].astype(str).str.strip()
    is_sepsis9 = diagnoses["icd_code"].isin(SEPSIS_ICD9) & (diagnoses["icd_version"] == 9)
    is_sepsis10 = (diagnoses["icd_version"] == 10) & diagnoses["icd_code"].str.startswith(SEPSIS_ICD10_PREFIXES)
    sepsis_hadm_ids = set(diagnoses.loc[is_sepsis9 | is_sepsis10, "hadm_id"].unique())

    comorbidity_count = diagnoses.groupby("hadm_id")["icd_code"].nunique().rename("comorbidity_count")

    cohort = icustays[icustays["hadm_id"].isin(sepsis_hadm_ids)].copy()
    cohort = cohort.merge(patients[["subject_id", "anchor_age", "anchor_year", "gender", "dod"]],
                           on="subject_id", how="left")
    cohort = cohort.merge(admissions[["hadm_id", "admittime", "deathtime"]], on="hadm_id", how="left")
    cohort = cohort.merge(comorbidity_count, on="hadm_id", how="left")

    cohort["intime"] = pd.to_datetime(cohort["intime"])
    cohort["outtime"] = pd.to_datetime(cohort["outtime"])
    cohort["admittime"] = pd.to_datetime(cohort["admittime"])
    cohort["age_at_admission"] = cohort["anchor_age"] + (cohort["admittime"].dt.year - cohort["anchor_year"])
    cohort = cohort[cohort["age_at_admission"] > 18].copy()

    # -- treatment: any vasopressor during the ICU stay --------------------
    vaso = con.execute(f"""
        select stay_id, starttime from read_csv_auto('{csv('icu/inputevents.csv.gz')}')
        where itemid in {VASOPRESSOR_ITEMIDS}
    """).df()
    treated_stays = set(vaso["stay_id"].unique())
    cohort["treatment"] = cohort["stay_id"].isin(treated_stays).astype(int)

    # -- baseline covariates: first lactate / first MAP during the stay ----
    lact = con.execute(f"""
        select subject_id, hadm_id, charttime, valuenum as lactate
        from read_csv_auto('{csv('hosp/labevents.csv.gz')}')
        where itemid in {LACTATE_ITEMIDS} and valuenum is not null
    """).df()
    lact["charttime"] = pd.to_datetime(lact["charttime"])
    lact = lact.merge(cohort[["subject_id", "hadm_id", "stay_id", "intime", "outtime"]],
                       on=["subject_id", "hadm_id"], how="inner")
    lact = lact[(lact["charttime"] >= lact["intime"]) & (lact["charttime"] <= lact["outtime"])]
    first_lactate = lact.sort_values("charttime").groupby("stay_id").first()["lactate"].rename("initial_lactate")

    mapv = con.execute(f"""
        select stay_id, charttime, valuenum as map_value
        from read_csv_auto('{csv('icu/chartevents.csv.gz')}')
        where itemid in {MAP_ITEMIDS} and valuenum is not null
    """).df()
    mapv["charttime"] = pd.to_datetime(mapv["charttime"])
    first_map = mapv.sort_values("charttime").groupby("stay_id").first()["map_value"].rename("initial_map")

    cohort = cohort.merge(first_lactate, on="stay_id", how="left")
    cohort = cohort.merge(first_map, on="stay_id", how="left")

    # -- outcome: 28-day mortality from ICU intime --------------------------
    cohort["dod"] = pd.to_datetime(cohort["dod"])
    cohort["deathtime"] = pd.to_datetime(cohort["deathtime"])
    death_time = cohort["deathtime"].combine_first(cohort["dod"])
    followup_end = cohort["intime"] + pd.Timedelta(days=FOLLOWUP_DAYS)
    died_in_window = death_time.notna() & (death_time <= followup_end)
    event_time = death_time.where(died_in_window, followup_end)
    cohort["time_days"] = (event_time - cohort["intime"]).dt.total_seconds() / 86400
    cohort["event_28d"] = died_in_window.astype(int)

    keep_cols = ["subject_id", "hadm_id", "stay_id", "age_at_admission", "gender",
                 "initial_lactate", "initial_map", "comorbidity_count",
                 "treatment", "time_days", "event_28d"]
    cohort = cohort[keep_cols].dropna(subset=["initial_lactate", "initial_map"]).reset_index(drop=True)
    cohort["gender_male"] = (cohort["gender"] == "M").astype(int)
    return cohort


# ---------------------------------------------------------------------------
# 2. Propensity score model + stabilized IPTW weights
# ---------------------------------------------------------------------------

COVARIATES = ["age_at_admission", "gender_male", "initial_lactate", "initial_map", "comorbidity_count"]


def fit_propensity_and_weights(cohort):
    X = sm.add_constant(cohort[COVARIATES])
    y = cohort["treatment"]
    model = sm.Logit(y, X).fit(disp=0)
    cohort["ps"] = model.predict(X)

    marginal_p = cohort["treatment"].mean()
    cohort["iptw"] = np.where(
        cohort["treatment"] == 1,
        marginal_p / cohort["ps"],
        (1 - marginal_p) / (1 - cohort["ps"]),
    )
    return cohort, model


# ---------------------------------------------------------------------------
# 3. Balance diagnostics: SMD before/after weighting
# ---------------------------------------------------------------------------

def smd_continuous(x, treat, weights=None):
    if weights is None:
        weights = np.ones(len(x))
    t, c = treat == 1, treat == 0
    m1 = np.average(x[t], weights=weights[t])
    m0 = np.average(x[c], weights=weights[c])
    v1 = np.average((x[t] - m1) ** 2, weights=weights[t])
    v0 = np.average((x[c] - m0) ** 2, weights=weights[c])
    return (m1 - m0) / np.sqrt((v1 + v0) / 2)


def balance_table(cohort):
    rows = []
    for cov in COVARIATES:
        unw = smd_continuous(cohort[cov].values, cohort["treatment"].values)
        w = smd_continuous(cohort[cov].values, cohort["treatment"].values, cohort["iptw"].values)
        rows.append({"covariate": cov, "smd_unweighted": unw, "smd_weighted": w})
    return pd.DataFrame(rows)


def love_plot(balance_df, path):
    fig, ax = plt.subplots(figsize=(6, 4))
    y = np.arange(len(balance_df))
    ax.scatter(balance_df["smd_unweighted"].abs(), y, label="Unweighted", color="tab:red")
    ax.scatter(balance_df["smd_weighted"].abs(), y, label="IPTW-weighted", color="tab:blue")
    ax.axvline(0.1, linestyle="--", color="gray", linewidth=1)
    ax.set_yticks(y)
    ax.set_yticklabels(balance_df["covariate"])
    ax.set_xlabel("Absolute Standardized Mean Difference")
    ax.set_title("Covariate balance before/after IPTW (pipeline demo)")
    ax.legend()
    fig.tight_layout()
    fig.savefig(path, dpi=150)
    plt.close(fig)


# ---------------------------------------------------------------------------
# 4. Weighted Cox model + weighted KM curve
# ---------------------------------------------------------------------------

def fit_weighted_cox(cohort):
    df = cohort[["time_days", "event_28d", "treatment", "iptw"]].copy()
    cph = CoxPHFitter()
    cph.fit(df, duration_col="time_days", event_col="event_28d",
            weights_col="iptw", robust=True)
    return cph


def weighted_km_plot(cohort, path):
    fig, ax = plt.subplots(figsize=(6, 4))
    kmf = KaplanMeierFitter()
    for label, grp in cohort.groupby("treatment"):
        name = "Early vasopressor" if label == 1 else "Standard care"
        kmf.fit(grp["time_days"], event_observed=grp["event_28d"],
                weights=grp["iptw"], label=name)
        kmf.plot_survival_function(ax=ax)
    ax.set_xlabel("Days from ICU admission")
    ax.set_ylabel("Survival probability")
    ax.set_title("IPTW-weighted Kaplan-Meier (pipeline demo — not a result)")
    fig.tight_layout()
    fig.savefig(path, dpi=150)
    plt.close(fig)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    con = duckdb.connect()
    cohort = build_cohort(con)
    print(f"Broadened demo cohort: n={len(cohort)}  "
          f"(treated={cohort['treatment'].sum()}, control={(cohort['treatment']==0).sum()})")

    if cohort["treatment"].nunique() < 2 or len(cohort) < 10:
        print("Not enough patients/variation in this demo pull to fit the pipeline. Stopping.")
        return

    cohort, ps_model = fit_propensity_and_weights(cohort)
    print("\nPropensity score model (logistic regression) summary:")
    print(ps_model.summary2().tables[1])

    balance_df = balance_table(cohort)
    print("\nBalance table (SMD before/after IPTW):")
    print(balance_df.to_string(index=False))
    love_plot(balance_df, REPORT_DIR / "love_plot_demo.png")

    cph = fit_weighted_cox(cohort)
    print("\nWeighted Cox model (treatment effect on 28-day mortality):")
    print(cph.summary[["coef", "exp(coef)", "exp(coef) lower 95%", "exp(coef) upper 95%", "p"]])

    weighted_km_plot(cohort, REPORT_DIR / "km_curve_demo.png")

    cohort.to_csv(OUT_DIR / "broadened_cohort_demo.csv", index=False)
    balance_df.to_csv(REPORT_DIR / "table1_balance_demo.csv", index=False)
    cph.summary.to_csv(REPORT_DIR / "table2_cox_demo.csv")

    print(f"\nOutputs written to {REPORT_DIR} and {OUT_DIR} (gitignored data; plots/tables are fine to commit).")
    print("\nREMINDER: this is a pipeline demonstration on a loosened, ~20-patient")
    print("subgroup of the 100-patient MIMIC-IV Demo. None of these numbers are a")
    print("scientific finding — they exist only to prove the code runs correctly")
    print("before being pointed at the real, credentialed MIMIC-IV cohort.")


if __name__ == "__main__":
    main()

File: analysis/pipeline_demo_broadened.R
# ---------------------------------------------------------------------------
# pipeline_demo_broadened.R
#
# PIPELINE DEMONSTRATION ONLY - NOT A SCIENTIFIC RESULT.
#
# Same broadened-cohort pipeline as the Python version (all sepsis ICU stays
# in the MIMIC-IV Demo, split by "received a vasopressor" vs. "did not" -
# NOT the strict Cryptic Shock definition). Exists to build/debug the
# propensity score, IPTW, balance diagnostics, and Cox model code ahead of
# full credentialed MIMIC-IV access. Do not report any output as a finding.
#
# Requires: readr, dplyr, lubridate, survival, ggplot2, tidyr, stringr
# ---------------------------------------------------------------------------

library(readr)
library(dplyr)
library(lubridate)
library(survival)
library(ggplot2)
library(tidyr)
library(stringr)

ROOT <- getwd()
# If not running in RStudio, just set this manually instead:
# ROOT <- "path/to/TTE_Cryptic_Shock_Project"

DATA_DIR   <- file.path(ROOT, "data", "mimic-iv-clinical-database-demo-2.2")
OUT_DIR    <- file.path(ROOT, "data", "processed")
REPORT_DIR <- file.path(ROOT, "report")
dir.create(OUT_DIR, recursive = TRUE, showWarnings = FALSE)
dir.create(REPORT_DIR, recursive = TRUE, showWarnings = FALSE)

LACTATE_ITEMIDS     <- c(50813, 52442, 53154)
MAP_ITEMIDS         <- c(220052, 220181)
VASOPRESSOR_ITEMIDS <- c(221906, 221289, 222315, 221749, 221662)
FOLLOWUP_DAYS       <- 28

SEPSIS_ICD9 <- c("0380","0381","03810","03811","03812","03819","0382","0383",
                 "03840","03841","03842","03843","03844","03849","0388","0389",
                 "99591","99592","78552")
SEPSIS_ICD10_PREFIXES <- c("A40", "A41", "R6520", "R6521")

read_gz <- function(relpath) {
  read_csv(file.path(DATA_DIR, relpath), show_col_types = FALSE)
}

# readr auto-detects date/datetime columns on read, so by the time a column
# reaches our code it may already be POSIXct/Date rather than character.
# Calling ymd_hms() directly on an already-parsed column can silently fail
# to parse (this is what caused "270 failed to parse" earlier). Wrapping in
# as.character() first, then using parse_date_time() with multiple possible
# formats, makes this robust regardless of what readr already did.
parse_dt <- function(x) {
  parse_date_time(as.character(x), orders = c("Ymd HMS", "Ymd HM", "Ymd"), quiet = TRUE)
}

# ---------------------------------------------------------------------------
# 1. Build the broadened cohort
# ---------------------------------------------------------------------------

build_cohort <- function() {
  patients   <- read_gz("hosp/patients.csv.gz")
  admissions <- read_gz("hosp/admissions.csv.gz")
  icustays   <- read_gz("icu/icustays.csv.gz")
  diagnoses  <- read_gz("hosp/diagnoses_icd.csv.gz") %>%
    mutate(icd_code = str_trim(as.character(icd_code)))
  
  is_sepsis9  <- diagnoses$icd_code %in% SEPSIS_ICD9 & diagnoses$icd_version == 9
  is_sepsis10 <- diagnoses$icd_version == 10 &
    str_starts(diagnoses$icd_code, paste(SEPSIS_ICD10_PREFIXES, collapse = "|"))
  sepsis_hadm_ids <- unique(diagnoses$hadm_id[is_sepsis9 | is_sepsis10])
  
  comorbidity_count <- diagnoses %>%
    group_by(hadm_id) %>%
    summarise(comorbidity_count = n_distinct(icd_code), .groups = "drop")
  
  cohort <- icustays %>%
    filter(hadm_id %in% sepsis_hadm_ids) %>%
    left_join(patients %>% select(subject_id, anchor_age, anchor_year, gender, dod),
              by = "subject_id") %>%
    left_join(admissions %>% select(hadm_id, admittime, deathtime), by = "hadm_id") %>%
    left_join(comorbidity_count, by = "hadm_id") %>%
    mutate(
      intime    = parse_dt(intime),
      outtime   = parse_dt(outtime),
      admittime = parse_dt(admittime),
      age_at_admission = anchor_age + (year(admittime) - anchor_year)
    ) %>%
    filter(age_at_admission > 18)
  
  # -- treatment: any vasopressor during the ICU stay -------------------
  vaso <- read_gz("icu/inputevents.csv.gz") %>%
    filter(itemid %in% VASOPRESSOR_ITEMIDS)
  treated_stays <- unique(vaso$stay_id)
  cohort <- cohort %>% mutate(treatment = as.integer(stay_id %in% treated_stays))
  
  # -- baseline covariates: first lactate / first MAP during the stay ---
  lact <- read_gz("hosp/labevents.csv.gz") %>%
    filter(itemid %in% LACTATE_ITEMIDS, !is.na(valuenum)) %>%
    rename(lactate = valuenum) %>%
    mutate(charttime = parse_dt(charttime)) %>%
    inner_join(cohort %>% select(subject_id, hadm_id, stay_id, intime, outtime),
               by = c("subject_id", "hadm_id"), relationship = "many-to-many") %>%
    filter(charttime >= intime, charttime <= outtime) %>%
    arrange(charttime) %>%
    group_by(stay_id) %>%
    slice(1) %>%
    ungroup() %>%
    select(stay_id, initial_lactate = lactate)
  
  mapv <- read_gz("icu/chartevents.csv.gz") %>%
    filter(itemid %in% MAP_ITEMIDS, !is.na(valuenum)) %>%
    rename(map_value = valuenum) %>%
    mutate(charttime = parse_dt(charttime)) %>%
    arrange(charttime) %>%
    group_by(stay_id) %>%
    slice(1) %>%
    ungroup() %>%
    select(stay_id, initial_map = map_value)
  
  cohort <- cohort %>%
    left_join(lact, by = "stay_id") %>%
    left_join(mapv, by = "stay_id")
  
  # -- outcome: 28-day mortality from ICU intime -------------------------
  cohort <- cohort %>%
    mutate(
      dod          = parse_dt(dod),
      deathtime    = parse_dt(deathtime),
      death_time   = coalesce(deathtime, dod),
      followup_end = intime + days(FOLLOWUP_DAYS),
      died_in_window = !is.na(death_time) & death_time <= followup_end,
      event_time   = if_else(died_in_window, death_time, followup_end),
      time_days    = as.numeric(difftime(event_time, intime, units = "days")),
      event_28d    = as.integer(died_in_window),
      gender_male  = as.integer(gender == "M")
    ) %>%
    select(subject_id, hadm_id, stay_id, age_at_admission, gender_male,
           initial_lactate, initial_map, comorbidity_count,
           treatment, time_days, event_28d) %>%
    drop_na(initial_lactate, initial_map)
  
  cohort
}

# ---------------------------------------------------------------------------
# 2. Propensity score model + stabilized IPTW weights
# ---------------------------------------------------------------------------

COVARIATES <- c("age_at_admission", "gender_male", "initial_lactate",
                "initial_map", "comorbidity_count")

fit_propensity_and_weights <- function(cohort) {
  form <- as.formula(paste("treatment ~", paste(COVARIATES, collapse = " + ")))
  ps_model <- glm(form, data = cohort, family = binomial())
  cohort$ps <- predict(ps_model, type = "response")
  
  marginal_p <- mean(cohort$treatment)
  cohort$iptw <- ifelse(
    cohort$treatment == 1,
    marginal_p / cohort$ps,
    (1 - marginal_p) / (1 - cohort$ps)
  )
  list(cohort = cohort, model = ps_model)
}

# ---------------------------------------------------------------------------
# 3. Balance diagnostics: SMD before/after weighting
# ---------------------------------------------------------------------------

weighted_var <- function(x, w, mean_x) {
  sum(w * (x - mean_x)^2) / sum(w)
}

smd_continuous <- function(x, treat, w = NULL) {
  if (is.null(w)) w <- rep(1, length(x))
  t_idx <- treat == 1
  c_idx <- treat == 0
  m1 <- weighted.mean(x[t_idx], w[t_idx])
  m0 <- weighted.mean(x[c_idx], w[c_idx])
  v1 <- weighted_var(x[t_idx], w[t_idx], m1)
  v0 <- weighted_var(x[c_idx], w[c_idx], m0)
  (m1 - m0) / sqrt((v1 + v0) / 2)
}

balance_table <- function(cohort) {
  rows <- lapply(COVARIATES, function(cov) {
    unw <- smd_continuous(cohort[[cov]], cohort$treatment)
    w   <- smd_continuous(cohort[[cov]], cohort$treatment, cohort$iptw)
    data.frame(covariate = cov, smd_unweighted = unw, smd_weighted = w)
  })
  bind_rows(rows)
}

love_plot <- function(balance_df, path) {
  plot_df <- balance_df %>%
    pivot_longer(cols = c(smd_unweighted, smd_weighted),
                 names_to = "type", values_to = "smd") %>%
    mutate(smd = abs(smd),
           type = recode(type, smd_unweighted = "Unweighted", smd_weighted = "IPTW-weighted"))
  
  p <- ggplot(plot_df, aes(x = smd, y = covariate, color = type)) +
    geom_point(size = 3) +
    geom_vline(xintercept = 0.1, linetype = "dashed", color = "gray40") +
    labs(x = "Absolute Standardized Mean Difference", y = NULL, color = NULL,
         title = "Covariate balance before/after IPTW (pipeline demo)") +
    theme_minimal()
  
  ggsave(path, p, width = 6, height = 4, dpi = 150)
}

# ---------------------------------------------------------------------------
# 4. Weighted Cox model + weighted KM curve
# ---------------------------------------------------------------------------

fit_weighted_cox <- function(cohort) {
  coxph(Surv(time_days, event_28d) ~ treatment, data = cohort,
        weights = iptw, robust = TRUE)
}

weighted_km_plot <- function(cohort, path) {
  fit <- survfit(Surv(time_days, event_28d) ~ treatment, data = cohort, weights = iptw)
  
  png(path, width = 6, height = 4, units = "in", res = 150)
  plot(fit, col = c("red", "blue"), lwd = 2,
       xlab = "Days from ICU admission", ylab = "Survival probability",
       main = "IPTW-weighted Kaplan-Meier (pipeline demo - not a result)")
  legend("bottomleft", legend = c("Standard care", "Early vasopressor"),
         col = c("red", "blue"), lwd = 2)
  dev.off()
}

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

main <- function() {
  cohort <- build_cohort()
  cat(sprintf("Broadened demo cohort: n=%d (treated=%d, control=%d)\n",
              nrow(cohort), sum(cohort$treatment), sum(cohort$treatment == 0)))
  
  if (n_distinct(cohort$treatment) < 2 || nrow(cohort) < 10) {
    cat("Not enough patients/variation in this demo pull to fit the pipeline. Stopping.\n")
    return(invisible(NULL))
  }
  
  res <- fit_propensity_and_weights(cohort)
  cohort <- res$cohort
  cat("\nPropensity score model (logistic regression) summary:\n")
  print(summary(res$model))
  
  balance_df <- balance_table(cohort)
  cat("\nBalance table (SMD before/after IPTW):\n")
  print(balance_df)
  love_plot(balance_df, file.path(REPORT_DIR, "love_plot_demo_R.png"))
  
  cox_model <- fit_weighted_cox(cohort)
  cat("\nWeighted Cox model (treatment effect on 28-day mortality):\n")
  print(summary(cox_model))
  
  weighted_km_plot(cohort, file.path(REPORT_DIR, "km_curve_demo_R.png"))
  
  write_csv(cohort, file.path(OUT_DIR, "broadened_cohort_demo_R.csv"))
  write_csv(balance_df, file.path(REPORT_DIR, "table1_balance_demo_R.csv"))
  write_csv(as.data.frame(summary(cox_model)$coefficients),
            file.path(REPORT_DIR, "table2_cox_demo_R.csv"))
  
  cat("\nREMINDER: this is a pipeline demonstration on a loosened, ~20-patient\n")
  cat("subgroup of the 100-patient MIMIC-IV Demo. None of these numbers are a\n")
  cat("scientific finding - they exist only to prove the code runs correctly\n")
  cat("before being pointed at the real, credentialed MIMIC-IV cohort.\n")
}

main()
